pkg/operators/oci/oci.go

```diff
@@ -0,0 +1,275 @@
+			DefaultValue: oci.DefaultAuthFile,
+			TypeHint:     params.TypeString,
+		},
+		{
+			Key:          validateMetadataParam,
+			Title:        "Validate metadata",
+			Description:  "Validate the gadget metadata before running the gadget",
+			DefaultValue: "true",
+			TypeHint:     params.TypeBool,
+		},
+		{
+			Key:          insecureParam,
+			Title:        "Insecure connection",
+			Description:  "Allow connections to HTTP only registries",
+			DefaultValue: "false",
+			TypeHint:     params.TypeBool,
+		},
+		{
+			Key:          pullParam,
+			Title:        "Pull policy",
+			Description:  "Specify when the gadget image should be pulled",
+			DefaultValue: oci.PullImageMissing,
+			PossibleValues: []string{
+				oci.PullImageAlways,
+				oci.PullImageMissing,
+				oci.PullImageNever,
+			},
+			TypeHint: params.TypeString,
+		},
+		{
+			Key:         pullSecret,
+			Title:       "Pull secret",
+			Description: "Secret to use when pulling the gadget image",
+			TypeHint:    params.TypeString,
+		},
+	}
+}
+
+func (o *ociHandler) Instantiate(gadgetCtx operators.GadgetContext, params *params.Params) (*OciHandlerInstance, error) {
+	if len(gadgetCtx.ImageName()) == 0 {
+		return nil, fmt.Errorf("imageName empty")
+	}
+
+	secretBytes := []byte{} // TODO
+
+	authOpts := &oci.AuthOptions{
+		AuthFile:    params.Get(authfileParam).AsString(),
+		SecretBytes: secretBytes,
+		Insecure:    params.Get(insecureParam).AsBool(),
+	}
+
+	// Make sure the image is available, either through pulling or by just accessing a local copy
+	// TODO: add security constraints (e.g. don't allow pulling - add GlobalParams for that)
+	err := oci.EnsureImage(gadgetCtx.Context(), gadgetCtx.ImageName(), authOpts, params.Get(pullParam).AsString())
+	if err != nil {
+		return nil, fmt.Errorf("insuring image: %w", err)
+	}
+
+	manifest, err := oci.GetManifestForHost(gadgetCtx.Context(), gadgetCtx.ImageName())
+	if err != nil {
+		return nil, fmt.Errorf("getting manifest: %w", err)
+	}
+
+	logger := gadgetCtx.Logger()
+	logger.Debugf("ArtifactType: %s", manifest.ArtifactType)
+
+	// metadata := &types.GadgetMetadata{}
+	r, err := oci.GetContentFromDescriptor(gadgetCtx.Context(), manifest.Config)
+	if err != nil {
+		return nil, fmt.Errorf("getting metadata: %w", err)
+	}
+	metadata, _ := io.ReadAll(r)
+	r.Close()
+
+	// Store metadata for serialization
+	gadgetCtx.SetMetadata(metadata)
+
+	viper := viper.New()
+	viper.SetConfigType("yaml")
+	err = viper.ReadConfig(bytes.NewReader(metadata))
+
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling metadata: %w", err)
+	}
+
+	gadgetCtx.SetVar("config", viper)
+
+	instance := &OciHandlerInstance{
+		gadgetCtx: gadgetCtx,
+	}
+
+	for _, layer := range manifest.Layers {
+		logger.Debugf("layer > %+v", layer)
+		ops := operators.GetImageOperatorsForMediaType(layer.MediaType)
+		for _, op := range ops {
+			logger.Debugf("found layer op %q", op.Name())
+			opInst, err := op.InstantiateImageOperator(gadgetCtx, layer)
```

Comment:
Perhaps instead of passing an interface, we can directly pass the contents of the layers the operator registered for. Something like `map[string][]byte`.

Reply:
That's another option.

Reply:
> Perhaps instead of passing an interface, we can directly pass the contents of the layers the operator registered for. Something like `map[string][]byte`.

Even better, we should pass an `io.Reader{Closer?}`. I think it'll help with modularization: an operator won't be able to access layers it's not registered for. 

I've also got a broader question on this topic: How should an operator register for multiple media types? I know it could call `operators.RegisterOperatorForMediaType` multiple times, but then `InstantiateImageOperator` will be called for each layer. Shouldn't be a single call to `InstantiateImageOperator` passing all layers it registered for? 

To add more context to this question, I found this out when implementing #2598, the logic to handle btfgen files could be added in the ebpf operator, in this case, it'll have to register for two different layers, but handling the logic in two different calls `InstantiateImageOperator` to seems too complicated to me.

Reply:
> Even better, we should pass an io.Reader{Closer?}. I think it'll help with modularization: an operator won't be able to access layers it's not registered for.
>
Added to https://github.com/inspektor-gadget/inspektor-gadget/issues/2653 to further discuss it.

> How should an operator register for multiple media types?
@mauriciovasquezbernal Do you have a use case in mind for this?